src/internal/operations/export.go

```diff
@@ -0,0 +1,423 @@
+
+	go func() {
+		defer writer.Close()
+		defer wr.Close()
+
+		for _, ec := range expCollections {
+			folder := ec.GetBasePath()
+			items := ec.GetItems(ctx)
+
+			for item := range items {
+				err := item.Error
+				if err != nil {
+					writer.CloseWithError(clues.Wrap(err, "getting export item").With("id", item.ID))
+					return
+				}
+
+				name := item.Data.Name
+
+				f, err := wr.Create(folder + "/" + name)
+				if err != nil {
+					writer.CloseWithError(clues.Wrap(err, "creating zip entry").With("name", name).With("id", item.ID))
+					return
+				}
+
+				_, err = io.Copy(f, item.Data.Body)
+				if err != nil {
+					writer.CloseWithError(clues.Wrap(err, "writing zip entry").With("name", name).With("id", item.ID))
+					return
+				}
+			}
+		}
+	}()
+
+	return zipExportCol{reader}, nil
+}
+
+func (op *ExportOperation) do(
+	ctx context.Context,
+	opStats *exportStats,
+	detailsStore streamstore.Reader,
+	start time.Time,
+) ([]export.Collection, error) {
+	logger.Ctx(ctx).
+		With("control_options", op.Options, "selectors", op.Selectors).
+		Info("restoring selection")
+
+	bup, deets, err := getBackupAndDetailsFromID(
+		ctx,
+		op.BackupID,
+		op.store,
+		detailsStore,
+		op.Errors)
+	if err != nil {
+		return nil, clues.Wrap(err, "getting backup and details")
+	}
+
+	observe.Message(ctx, "Restoring", observe.Bullet, clues.Hide(bup.Selector.DiscreteOwner))
+
+	paths, err := formatDetailsForRestoration(ctx, bup.Version, op.Selectors, deets, op.Errors)
+	if err != nil {
+		return nil, clues.Wrap(err, "formatting paths from details")
+	}
+
+	ctx = clues.Add(
+		ctx,
+		"resource_owner_id", bup.Selector.ID(),
+		"resource_owner_name", clues.Hide(bup.Selector.Name()),
+		"details_entries", len(deets.Entries),
+		"details_paths", len(paths),
+		"backup_snapshot_id", bup.SnapshotID,
+		"backup_version", bup.Version)
+
+	op.bus.Event(
+		ctx,
+		events.ExportStart,
+		map[string]any{
+			events.StartTime:        start,
+			events.BackupID:         op.BackupID,
+			events.BackupCreateTime: bup.CreationTime,
+			events.ExportID:         opStats.exportID,
+		})
+
+	observe.Message(ctx, fmt.Sprintf("Discovered %d items in backup %s to export", len(paths), op.BackupID))
+
+	kopiaComplete := observe.MessageWithCompletion(ctx, "Enumerating items in repository")
+	defer close(kopiaComplete)
+
+	dcs, err := op.kopia.ProduceRestoreCollections(ctx, bup.SnapshotID, paths, opStats.bytesRead, op.Errors)
+	if err != nil {
+		return nil, clues.Wrap(err, "producing collections to export")
+	}
+
+	kopiaComplete <- struct{}{}
+
+	ctx = clues.Add(ctx, "coll_count", len(dcs))
+
+	// should always be 1, since backups are 1:1 with resourceOwners.
+	opStats.resourceCount = 1
+	opStats.cs = dcs
+
+	expCollections, err := exportRestoreCollections(
+		ctx,
+		op.ec,
+		bup.Version,
+		op.Selectors,
+		op.ExportCfg,
+		op.Options,
+		dcs,
+		op.Errors)
+	if err != nil {
+		return nil, clues.Wrap(err, "restoring collections")
+	}
+
+	opStats.ctrl = op.ec.Wait()
+
```

Comment:
I'm not completely up to date on all the m365 package stuff. Is this actually adding anything to the implementation? I thought it was to make sure we emptied all the collections the m365 package makes during backup?

Reply:
Nothing much besides gathering metrics. I was mostly just mimicking what restore was doing.